Cpu resources

skinowski opened the pull request:
As discussed in #405 , this change adds CPUs param to routes and calls. This is represented by a string in ether of the following formats:

*) as MilliCPUs eg. 100m which is similar to k8s which means 1/10 of a CPU core or
*) as floating-point eg. 0.5 which means 1/2 of a CPU core, similar to docker 'cpus'

Resource allocation for CPU is handled very similar to Memory except during initialization there's no headroom estimation for fn server and container run time (docker) cpu consumption itself.

Internally, CPUs is converted to millicpu units to set CFS period/quota via docker client API.

Mixed workloads (unlimited CPUs calls mixed with non-zero CPUs calls) is left open, so operators have to keep an eye on that either by not allowing zero CPUs calls at all or only allowing zero CPUs calls to disable CPU resource allocation.

denismakogon commented on api/agent/agent_test.go:
```diff
@@ -412,6 +422,7 @@ func TestSubmitError(t *testing.T) {
 		"FN_APP_NAME": appName,
 		"FN_PATH":     path,
 		"FN_MEMORY":   strconv.Itoa(memory),
+		"FN_CPUS":     CPUs,
```
Overall looks good, but i do feel like there would be a problem with UX. How would user know the capacity for CPU and RAM? I do feel like there should be some sort of API to see the exact free/used capacity (think of flavors on AWS, OpenStack, etc.).

skinowski replied:
Yes, I need to add these as stats (CPU, mem, slots in use, hot/cold container counts, etc. etc.) and maybe a API could also be useful. I'll open an issue to discuss and proceed.

rdallman commented on api/agent/resource.go:
```diff
@@ -68,78 +80,93 @@ func (t *resourceToken) Close() error {
 
 		t := &resourceToken{decrement: func() {
+
+			// Guard against multiple close
```
wrapping the `.decrement()` call in https://golang.org/pkg/sync/#Once.Do inside of `.Close()` may be a little cleaner than using a variable from the closure's enclosing scope

rdallman commented on api/agent/resource.go:
```diff
@@ -286,24 +378,59 @@ func getMemoryHeadRoom(usableMemory uint64) (uint64, error) {
 
-func checkCgroup() (uint64, error) {
-	f, err := os.Open("/sys/fs/cgroup/memory/memory.limit_in_bytes")
+func readString(fileName string) (string, error) {
```
~dupe code: https://golang.org/pkg/io/ioutil/#ReadFile

rdallman commented on api/models/route.go:
```diff
@@ -128,9 +133,52 @@ func (r *Route) Validate() error {
+	// was submitted.
+	tmp, err := strconv.ParseFloat(sanitized, 64)
+	if err != nil || cpu != tmp {
+		return "", ErrRoutesInvalidCPUs
```
can we just floor this? the error message doesn't make it clear that they must specify only up to 2 decimals precision, and it doesn't seem to warrant an error really to lop off any more precision. up to you, but need to make error clearly state what's wrong or floor / ceiling I think.

skinowski replied:
I was thinking about adding this to the documentation where expected format is 2 decimal points precision. We kind of do the same for other settings, eg. we don't have ErrMemoryTooLarge or TimeoutTooBig, so I stuck with one error code. I could round it silently too. Not sure.

rdallman replied:
I think that's fine, just the error could be more specific about its format / bounds so that users could debug it quicker than having to go look up the docs (we'll get a slack message long before then). rounding for this seems much easier and pretty inconsequential.

skinowski commented on api/agent/call.go:
```diff
@@ -392,6 +407,21 @@ func (c *call) End(ctx context.Context, errIn error) error {
+// getCPUQuota converts CPUs to CPU quota by multiplying them by 100.
+// eg. 0.20 becomes 20, 2.0 becomes 200, which are easier to calculate
+// and manage as resources and translate to Linux CFS (cpu-quota and cpu-period)
+func getCPUQuota(CPUs string) (uint64, error) {
```
So, the flow is more like string -> float -> int. I opted for string since we may extend it to k8s style milicpus, etc. formats. We end up converting all of this to uint64 at driver level:

https://godoc.org/github.com/moby/moby/api/types/container#Resources

Maybe we should stick with uint64 from the front and stay consistent using millicpu, where "100" means 100% (1CPU), etc.

skinowski replied:
Assuming we are ok with "string -> float -> int" flow, I'm going to clean up (and clarify this) further in the next commit.

rdallman replied:
it would be nice if it were either a float or an int, and then the type for it can add `UnmarshalJSON`, etc., if we want the format to be `"cpus":"100mCPU"` or something, it would be pretty flexible. i think we should just pick something, though, instead of having a string we have to parse out much later down the line. then the user is confronted with their poor choices when they're adding it to the route and there's no chance of a mishap (error) in the backend, we can just use the thing.

i.e. there'd be no string->float->int, just parse to an int or float immediately, and then the driver can do what it needs to with that. it's a little strange to see a number in a string, as it is, in a json api. I'd be okay with just picking 'milliCPUs' to be consistent with k8s now and calling it a day. if we end up regretting that it should be easy to work around, but I doubt it.

skinowski commented:
So, this version is interesting..

I've added the MilliCPUs type, but to fetch proper API errors, I had add a few checks every time we do json handling (in order to return specific error code and not generic Invalid JSON)

I ran into some dependency chaos, so I chose not to use MilliCPUs inside docker driver. And resource.go has been unrelated to models so far, so I kept it as is and ended up doing a few uint64 casts.

In DB, I modified this into 'int' column similar to memory, timeouts.

This handles float format which is 0.2 => 20% of 1 CPU as well as k8s millicpus, which is 1000m => 1 CPU, etc.  MilliCPUs have a suffix of 'm', similar to k8s.

rdallman commented:
> I've added the MilliCPUs type, but to fetch proper API errors, I had add a few checks every time we do json handling (in order to return specific error code and not generic Invalid JSON)

ah, crap that's right gin isn't very flexible in this dept. we should probably have our own json helper function instead of using gin's in handlers (that uses gin's, sure) since we need to return our own error type if things go awry.

rdallman commented on api/models/config.go:
```diff
@@ -125,3 +127,67 @@ func (h *Headers) Scan(value interface{}) error {
+// implements json.Unmarshaler
+func (c *MilliCPUs) UnmarshalJSON(data []byte) error {
+
+	outer := strings.TrimSpace(string(data))
```
not a huge deal, but useful to know, golang `strings` and `bytes` package have nearly identical helper functions for trim/suffix/prefix/etc work. https://golang.org/pkg/bytes/#TrimSpace

rdallman commented on api/models/config.go:
```diff
@@ -125,3 +127,67 @@ func (h *Headers) Scan(value interface{}) error {
+// implements json.Marshaler
+func (c *MilliCPUs) MarshalJSON() ([]byte, error) {
+
+	if *c < MinMilliCPUs || *c > MaxMilliCPUs {
```
kind of pedantic but this check here and in unmarshal seem like they might belong in `route.Validate()` since they're not parsing errors. this might get rid of the noted json handling woes

rdallman replied:
i guess not quite on the json woes, since they could still not provide a valid float / millicpu value

rdallman commented on api/models/error.go:
```diff
@@ -206,6 +207,11 @@ func (e err) Code() int { return e.code }
 func NewAPIError(code int, e error) APIError { return err{code, e} }
 
+func IsAPIError(e error) bool {
+	_, ok := e.(err)
```
nit: probably should be the interface. `_, ok := e.(APIError)`

rdallman commented:
> I ran into some dependency chaos, so I chose not to use MilliCPUs inside docker driver. And resource.go has been unrelated to models so far, so I kept it as is and ended up doing a few uint64 casts.

seems like a sane boundary 👍

skinowski commented:
@rdallman bump. :-)
